python/sparkdl/graph_builder.py

```diff
@@ -0,0 +1,329 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import joblib as jl
+import logging
+from pathlib import Path
+import shutil
+import six
+from tempfile import mkdtemp
+import webbrowser
+
+import keras.backend as K
+import tensorflow as tf
+import tensorframes.core as tfrm
+
+from .utils import JVMAPI
+
+logger = logging.getLogger('sparkdl')
+
+class GraphBuilder(object):
+    """ 
+    Building TensorFlow graph and export as either UDF or GraphFunction
+    
+    This is a thin layer on top of `tf.Session`.
+    """
+
+    def __init__(self, graph):
+        assert isinstance(graph, tf.Graph)
+        self.sess = tf.Session(graph=graph)
+        self.graph = graph
+
+    def get_shape(self, tnsr_or_name):
+        _shape = self.get_tensor(tnsr_or_name).get_shape().as_list()
+        return [-1 if x is None else x for x in _shape]
+    
+    def get_op(self, tfobj_or_name):
+        if isinstance(tfobj_or_name, tf.Operation):
+            return tfobj_or_name
+        name = tfobj_or_name
+        if isinstance(tfobj_or_name, tf.Tensor):
+            name = tfobj_or_name.name
+        if not isinstance(name, six.string_types):
+            raise TypeError('invalid op request for {} of {}'.format(name, type(name)))
+        name = name.split(":")[0]
+        op = self.graph.get_operation_by_name(name)
+        assert op is not None, \
+            'cannot locate op {} in current graph'.format(name)
+        return op        
+
+    def get_tensor(self, tfobj_or_name):
+        if isinstance(tfobj_or_name, tf.Tensor):
+            return tfobj_or_name
+        name = tfobj_or_name
+        if isinstance(tfobj_or_name, tf.Operation):
+            name = tfobj_or_name.name
+        if not isinstance(name, six.string_types):
+            raise TypeError('invalid tensor request for {} of {}'.format(name, type(name)))
+        name_parts = name.split(":")
+        if len(name_parts) < 2:
+            name += ":0"
+        tnsr = self.graph.get_tensor_by_name(name)
+        assert tnsr is not None, \
+            'cannot locate tensor {} in current graph'.format(name)
+        return tnsr
+
+    def op_name(self, tfobj_or_name):
+        return self.get_op(tfobj_or_name).name
+
+    def tensor_name(self, tfobj_or_name):
+        return self.get_tensor(tfobj_or_name).name
+
+    def valid_output(self, tfobj_or_name):
+        return self.op_name(tfobj_or_name)
+
+    def valid_input(self, tfobj_or_name):
+        name = self.op_name(tfobj_or_name)
+        op = self.graph.get_operation_by_name(name)
+        assert 'Placeholder' == op.type, \
+            ('input must be Placeholder, but get', op.type)
+        return name
+
+    def show_tf_graph(self, max_const_size=32):
```

Comment:
ok so this writes out a html file so that you can view the graph in your browser. it's awesome! let's rename it to write_visualization_html() and have it take in an optional filename